sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero_common.hpp

```diff
@@ -23,5 +28,158 @@
+// ready. Recognized values (can be used as a bit mask):
+enum {
+  ZeSerializeNone =
+      0, // no locking or blocking (except when SYCL RT requested blocking)
+  ZeSerializeLock = 1, // locking around each ZE_CALL
+  ZeSerializeBlock =
+      2, // blocking ZE calls, where supported (usually in enqueue commands)
+};
+
+static const uint32_t UrL0Serialize = [] {
+  const char *ZeSerializeMode = std::getenv("ZE_SERIALIZE");
+  const char *UrL0SerializeMode = std::getenv("UR_L0_SERIALIZE");
+  uint32_t SerializeModeValue = 0;
+  if (ZeSerializeMode) {
+    SerializeModeValue = std::atoi(ZeSerializeMode);
+  } else if (UrL0SerializeMode) {
+    SerializeModeValue = std::atoi(UrL0SerializeMode);
+  }
+  return SerializeModeValue;
+}();
+
+const int ZeSerialize = UrL0Debug;
+
+// This class encapsulates actions taken along with a call to Level Zero API.
+class ZeCall {
+private:
+  // The global mutex that is used for total serialization of Level Zero calls.
+  static std::mutex GlobalLock;
+
+public:
+  ZeCall() {
+    if ((UrL0Serialize & UrL0SerializeLock) != 0) {
+      GlobalLock.lock();
+    }
+  }
+  ~ZeCall() {
+    if ((UrL0Serialize & UrL0SerializeLock) != 0) {
+      GlobalLock.unlock();
+    }
+  }
+
+  // The non-static version just calls static one.
+  ze_result_t doCall(ze_result_t ZeResult, const char *ZeName,
+                     const char *ZeArgs, bool TraceError = true);
+};
+
+// Prints to stderr if ZE_DEBUG allows it
+void zePrint(const char *Format, ...);
+
+// This function will ensure compatibility with both Linux and Windows for
+// setting environment variables.
+bool setEnvVar(const char *name, const char *value);
+
+// Prints to stderr if UR_L0_DEBUG allows it
+void urPrint(const char *Format, ...);
+
+// Helper for one-liner validation
+#define UR_ASSERT(condition, error)                                            \
+  if (!(condition))                                                            \
+    return error;
+
+// Returns the ze_structure_type_t to use in .stype of a structured descriptor.
+// Intentionally not defined; will give an error if no proper specialization
+template <class T> ze_structure_type_t getZeStructureType();
+template <class T> zes_structure_type_t getZesStructureType();
+
+// The helpers to properly default initialize Level-Zero descriptor and
+// properties structures.
+template <class T> struct ZeStruct : public T {
+  ZeStruct() : T{} { // zero initializes base struct
+    this->stype = getZeStructureType<T>();
+    this->pNext = nullptr;
+  }
+};
+
+template <class T> struct ZesStruct : public T {
+  ZesStruct() : T{} { // zero initializes base struct
+    this->stype = getZesStructureType<T>();
+    this->pNext = nullptr;
+  }
+};
+
 // Map Level Zero runtime error code to UR error code.
 ur_result_t ze2urResult(ze_result_t ZeResult);
+
+// Trace a call to Level-Zero RT
+#define ZE2UR_CALL(ZeName, ZeArgs)                                             \
```

Comment:
Why does it not replace ZE_CALL?

Reply:
@smaslov-intel It does replace it, in a few patches after.